### PR TITLE
chore(broker-core): migrate membership management to atomix

### LINKED ISSUE #1924: Replace gossip with atomix
- listen to cluster ship changes from atomix
- register end points with transport (json serialization over properties) https://github.com/zeebe-io/zeebe/blob/81afb84cf1703cf47cf7781e85fb7669cda0a66e/broker-core/src/main/java/io/zeebe/broker/clustering/base/connections/RemoteAddressManager.java#L66-L73
- propagate raft changes (LEADER vs. FOLLOWER) and update topology https://github.com/zeebe-io/zeebe/blob/81afb84cf1703cf47cf7781e85fb7669cda0a66e/broker-core/src/main/java/io/zeebe/broker/clustering/base/topology/TopologyManagerImpl.java#L316-L332

### PR DESCRIPTION
**Pull request by deepthidevaki**
closes #1924

### CODE REVIEW

File: broker-core/src/main/java/io/zeebe/broker/clustering/ClusterComponent.java [hunk 1]
```diff
@@ -106,29 +103,15 @@ private void initGossip(
...
-        .install();
 
     final AtomixService atomixService = new AtomixService(context.getBrokerConfiguration());
-
     baseLayerInstall.createService(ATOMIX_SERVICE, atomixService).install();
+
+    final AtomixJoinService atomixJoinService = new AtomixJoinService();
+    baseLayerInstall
+        .createService(ATOMIX_JOIN_SERVICE, atomixJoinService)
+        .dependency(ATOMIX_SERVICE, atomixJoinService.getAtomixInjector())
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/clustering/ClusterComponent.java]: Can we have also a dependency on `TOPOLOGY_MANAGER_SERVICE`, so we can be sure that they start in order. I think this is also a bug in the old implementation.

File: broker-core/src/main/java/io/zeebe/broker/clustering/base/gossip/AtomixService.java [hunk 2]
```diff
@@ -84,16 +81,10 @@ public void start(ServiceStartContext startContext) {
...
-    final CompletableFuture<Void> startFuture = atomix.start();
-    startContext.async(mapCompletableFuture(startFuture));
   }
 
   @Override
-  public void stop(ServiceStopContext stopContext) {
-    final CompletableFuture<Void> stopFuture = atomix.stop();
-    stopContext.async(mapCompletableFuture(stopFuture));
-  }
+  public void stop(ServiceStopContext stopContext) {}
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/clustering/base/gossip/AtomixService.java]: we can remove this

File: broker-core/src/main/java/io/zeebe/broker/clustering/base/topology/TopologyManagerImpl.java [hunk 3]
```diff
@@ -84,23 +71,10 @@ public String getName() {
...
-  protected void onActorStarting() {
-    gossip.addMembershipListener(membershipListner);
-
-    gossip.addCustomEventListener(CONTACT_POINTS_EVENT_TYPE, contactPointsChangeListener);
-    gossip.addCustomEventListener(PARTITIONS_EVENT_TYPE, partitionChangeListener);
-
-    gossip.registerSyncRequestHandler(CONTACT_POINTS_EVENT_TYPE, localContactPointsSycHandler);
-    gossip.registerSyncRequestHandler(PARTITIONS_EVENT_TYPE, knownPartitionsSyncHandler);
-  }
+  protected void onActorStarting() {}
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/clustering/base/topology/TopologyManagerImpl.java]: Please remove the empty methods :)

File: broker-core/src/main/java/io/zeebe/broker/clustering/base/topology/TopologyManagerImpl.java [hunk 4]
```diff
@@ -145,104 +119,100 @@ public void onStateChange(Raft raft, RaftState raftState) {
...
+          switch (clusterMembershipEvent.type()) {
+            case METADATA_CHANGED:
+              LOG.debug(
+                  "Member {} receives metadata change of member {}", localNode, eventSource.id());
+              updatePartitionInfo(eventSource);
+              break;
+            case MEMBER_ADDED:
+              LOG.debug("Member {} receives event member {} added", localNode, eventSource.id());
+
+              final Properties newProperties = eventSource.properties();
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/clustering/base/topology/TopologyManagerImpl.java]: Can we extract this in a new method so this method is not that long, please?

File: broker-core/src/main/java/io/zeebe/broker/clustering/base/topology/TopologyManagerImpl.java [hunk 5]
```diff
@@ -145,104 +119,100 @@ public void onStateChange(Raft raft, RaftState raftState) {
...
+              final String replicationAddress = newProperties.getProperty("replicationAddress");
+              final String managementAddress = newProperties.getProperty("managementAddress");
+              final String clientApiAddress = newProperties.getProperty("clientAddress");
+              final String subscriptionAddress = newProperties.getProperty("subscriptionAddress");
+              final ObjectMapper mapper = new ObjectMapper();
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/clustering/base/topology/TopologyManagerImpl.java]: We can reuse the ObjectMapper in this class, simply create one on creation time.

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Looks good, just some small requests :+1:

**Comment by Zelldon** [comment]
Hey guys,

had over the weekend a closer look why the IT tests are failing and at our atomix setup. You can take a 
look at atomix-gossip-test branch for the latest changes.

As @deepthidevaki found out that PartitionGroup creating only on the first node is problematic on start up it is also error prone on closing the fist node. See this bug issue https://github.com/atomix/atomix/issues/999 , which I opend yesterday. That was an issue which we had in the IT tests.

We should also not use the `Atomix` class, instead more his base class `AtomixCluster`.
The `AtomixCluster` can be used without partition groups and is more for cluster bootstrapping, cluster membership and communication, which is currently our use case. If you want to use primitives on top, then you can use the `Atomix` class. This will reduce our memory and threading footprint for now as well.

Setup would look similar then before, but without the partition groups. I also configured the SwimProtocol, such that the leave time out is 2 sec instead of 10 (default). In some test we wait until a member is removed from the topology, which will only happen after the `MEMBER_REMOVED` is published. 

Currently on the atomix-gossip-test I have still failing tests, because some ports are not freed.

**Comment by deepthidevaki** [comment]
Thanks @zell for looking into the issue. We can incorporate your findings in a separate PR. I am finishing up this PR. @menski You can have a look at this.

**Comment by menski** [comment]
@deepthidevaki this change introduces some info log messages, do you thing they are useful and should be logged on info level? Maybe some should be removed, trace or debug? 

Also the message `Im node {}, event received from {} {}. Nothing to do` should be reworded in my opinion, or maybe just removed?

**Pull request merged**